src/operator/math_functions-inl.h

```diff
@@ -35,47 +35,29 @@ namespace op {
 namespace math {
 
 // Wrappers for math.h unary and binary functions
-// - For DType != double: math::name(a) does computation in float,
-//   casting back to DType afterwards
+// - For DType != double: math::name(a) does computation in float
+//   and returns float
 // - For DType == double: math::name(a) does computation in double
-// - math::namef(a) does computation in float and returns float value
-//   (no casting back). Use this in combined expressions (for example,
-//   math::log1p(math::expf(a)) in softrelu (mshadow_op.h))
+//   and returns double
 
 #define MXNET_UNARY_MATH_FUNC(name) \
 template<typename DType> MSHADOW_XINLINE \
-DType name(DType a) { \
-  return DType(::name##f(static_cast<float>(a))); \
-} \
-template<> MSHADOW_XINLINE \
-float name(float a) { \
-  return ::name##f(a); \
+float name(DType a) { \
+  return ::name##f(static_cast<float>(a)); \
 } \
-template<> MSHADOW_XINLINE \
+MSHADOW_XINLINE \
 double name(double a) { \
   return ::name(a); \
-} \
-template<typename DType> MSHADOW_XINLINE \
-float name##f(DType a) { \
-  return ::name##f(static_cast<float>(a)); \
 }
 
 #define MXNET_BINARY_MATH_FUNC(name) \
 template<typename DType> MSHADOW_XINLINE \
-DType name(DType a, DType b) { \
-  return DType(::name##f(static_cast<float>(a), static_cast<float>(b))); \
+float name(DType a, DType b) { \
```

Comment:
I probably don't understand the purpose of this PR, but is returning float for all DTypes intentional? That seems bad to me.

Reply:
Hello, if you look at mshadow_op.h, and in particular MXNET_UNARY_MATH_OP, we cast back to DType there.

This is all about math functions, for which we only have implementations in float and double. If you happen to have code for computing (say) exp in half_t natively, you'd have to overwrite the op in mshadow_op.h.

This is pretty much what was done before. I just changed it, so that for double, we compute things in double.

Reply:
I get the idea of we may have been casting double to float to pass to math functions that take a float as an argument.  I am missing how that translates into returning 'float' for any DType that isn't explicitly overloaded.  Why does this return float?

Reply:
The reason that math_functions-inl.h return float for all DType != double, is that they are being used in combined expressions, for example check sigmoid in mshadow_op.h. If the functions returned DType, you'd be casting back and forth and lose accuracy. You only want to cast back to DType in the very end.

This is something that @piiswrong fixed quite nicely.